Reduce use of memcpy/memset in CryptoAlgorithmAESCFBOpenSSL.cpp

AlexsanderDamaceno opened the pull request:
#### 224847f7031579dd9d72a27dc8abf70bffcd0535
<pre>
Reduce use of memcpy/memset in CryptoAlgorithmAESCFBOpenSSL.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=290828">https://bugs.webkit.org/show_bug.cgi?id=290828</a>

Reviewed by Geoffrey Garen.

Reduce use of memcpy/memset.
memcpy/memset is unsafe and can lead to memory corruption bugs

* Source/WebCore/crypto/openssl/CryptoAlgorithmAESCFBOpenSSL.cpp:
(WebCore::cfb8):

Canonical link: <a href="https://commits.webkit.org/293117@main">https://commits.webkit.org/293117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c879abd493eccf830e76b47f166a1815fc13906

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97856 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102969 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48385 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99901 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74506 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31684 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13458 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88436 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54863 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13234 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6363 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47828 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83222 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6436 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105348 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24936 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18160 "Found 1 new test failure: http/tests/pdf/linearized-pdf-in-display-none-iframe.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83520 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25308 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84612 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82956 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27573 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5262 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18545 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15863 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24897 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30066 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24719 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28033 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26293 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/crypto/openssl/CryptoAlgorithmAESCFBOpenSSL.cpp:
```diff
@@ -63,11 +65,11 @@ static std::optional<Vector<uint8_t>> cfb8(const Vector<uint8_t>& key, const Vec
         }
     }
-    memset(encryptedBlock, 0, sizeof encryptedBlock);
+    zeroSpan(std::span { encryptedBlock });
```
Are you sure you need the explicit `std::span` here? Implicit conversion should work.

AlexsanderDamaceno replied:
Unfortunately, it do not compile for some builds using implicit conversion

cdumez commented on Source/WebCore/crypto/openssl/CryptoAlgorithmAESCFBOpenSSL.cpp:
```diff
@@ -45,15 +46,16 @@ static std::optional<Vector<uint8_t>> cfb8(const Vector<uint8_t>& key, const Vec
-    uint8_t shiftRegister[AES_BLOCK_SIZE * 2];
-    memcpy(shiftRegister, iv.data(), AES_BLOCK_SIZE);
+    std::array<uint8_t, AES_BLOCK_SIZE * 2> shiftRegister;
+    auto shiftRegisterSpan = std::span(shiftRegister);
```
We can just write:
```
std::span shiftRegisterSpan { shiftRegister };
```

AlexsanderDamaceno committed:
Reduce use of memcpy/memset in CryptoAlgorithmAESCFBOpenSSL.cpp
https://bugs.webkit.org/show_bug.cgi?id=290828

Reviewed by Geoffrey Garen.

Reduce use of memcpy/memset.
memcpy/memset is unsafe and can lead to memory corruption bugs

* Source/WebCore/crypto/openssl/CryptoAlgorithmAESCFBOpenSSL.cpp:
(WebCore::cfb8):

Canonical link: https://commits.webkit.org/293117@main
